docs: add `Modal` API decisions (#1187)
razorpay/blade: packages/blade/src/components/Modal/_decisions/decisions.md

This document outlines the API of `Modal` component.

<!-- TODO: Add this -->
<!-- <img src="./modal-thumbnail.png" width="380" alt="Thumbnail of modal" /> -->

- [Design](#design)
- [Anatomy](#anatomy)
- [Animation](#animation)
- [API](#api)
  - [`Modal` API](#modal-api)
  - [`ModalBody` API](#modalbody-api)
  - [`ModalHeader` API](#modalheader-api)
  - [`ModalFooter` API](#modalfooter-api)
- [Behaviors](#behaviors)
  - [Modal Stacking](#modal-stacking)
  - [Modal Sizes \& Responsiveness](#modal-sizes--responsiveness)
  - [Modal Height](#modal-height)
  - [Modal Usage on mWeb \& native apps](#modal-usage-on-mweb--native-apps)
  - [Modal for React Native](#modal-for-react-native)
- [Accessibility](#accessibility)
- [Open questions](#open-questions)
- [References](#references)

## Design

- [Figma - Modal](https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?type=design&node-id=35860-607833&t=hEZsxB4yG6A3Gmib-0)

## Anatomy
<!-- TODO: Add a better screenshot -->
<img src="./modal-anatomy.png" alt="Modal Anatomy" width="100%" />

## Animation

> TBD

Author: Waiting on design to share this

- [API](#api)
  - [`Modal` API](#modal-api)
  - [`ModalBody` API](#modalbody-api)
  - [`ModalHeader` API](#modalheader-api)
  - [`ModalFooter` API](#modalfooter-api)
- [Behaviors](#behaviors)
  - [Modal Stacking](#modal-stacking)
  - [Modal Sizes \& Responsiveness](#modal-sizes--responsiveness)
  - [Modal Height](#modal-height)
  - [Modal Usage on mWeb \& native apps](#modal-usage-on-mweb--native-apps)
  - [Modal for React Native](#modal-for-react-native)
- [Accessibility](#accessibility)
- [Open questions](#open-questions)
- [References](#references)

## Design

- [Figma - Modal](https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?type=design&node-id=35860-607833&t=hEZsxB4yG6A3Gmib-0)

## Anatomy
<!-- TODO: Add a better screenshot -->
<img src="./modal-anatomy.png" alt="Modal Anatomy" width="100%" />

## Animation
**Entry:**
<img src="./modal-entry.gif" alt="Modal Entry Animation" width="100%" />

**Exit:**
<img src="./modal-exit.gif" alt="Modal Exit Animation" width="100%" />

## API

Sample usage:

```jsx
import { Modal, ModalHeader, ModalBody, ModalFooter, Button, Text } from '@razorpay/blade/components';

const [isOpen, setIsOpen] = useState(false);

> <Modal isOpen={isOpen} onDismiss={() => setIsOpen(false)}>

Reviewer: learnability of API++ 🚀

> `ModalFooter` API will be similar to [`BottomSheetFooter` API](https://github.com/razorpay/blade/blob/master/packages/blade/src/components/BottomSheet/_decisions/decisions.md#bottomsheetfooter-api).

| Prop       | Type              | Default     | Description                 | Required |
| ---------- | ----------------- | ----------- | --------------------------- | -------- |
| `children` | `React.ReactNode` | `undefined` | Contents of the ModalFooter | ✅       |

## Behaviors
### Modal Stacking
- We're at consensus that we do need to support modal stacking.
- Design is not yet aligned on how the modal should look like when stacked and what the motion should be.
- As an interim solution we can implement a basic stacking behavior with the same exit/entry animations as the current modal and look into enhancing this motion later.

### Modal Sizes & Responsiveness
- We will offer Modal in 3 sizes - small, medium and large.
- Each size corresponds to a max-width for the Modal
- As of today these are the max-widths we're using for each size:
  - small: `400px`
  - medium: `760px`
  - large: `1024px`
- All the Modals will have a min-width of `320px` regardless of the size.

> - This would mean that the Modal will be responsive and will scale down to `320px` on smaller screens but will not scale up beyond the max-width as per its size mentioned above.

Reviewer: Let's add a link to your POC for visualization.

Reviewer 2: add screenshot as well explaining this?

### Modal Height
- Each Modal will have a max-height of `80vh` (80% of viewport height) and will be scrollable if the content exceeds this height.

### Modal Usage on mWeb & native apps
- We will not be using the Modal component on mWeb or native apps.
- We will only use the BottomSheet component on mWeb and native apps.
- Our Modal will be responsive and will not break up till 320px on mWeb but will not be usable on native apps.

> - We will throw a warning when a Modal is opened on smaller screens.

Reviewer: TODO: document these on the modal storybook.

- Designers would need to be mindful of this while designing for mWeb and native apps.
- On Figma, we have added guardrails to ensure that when screenSize is selected Mobile on Figma, Modal component will turn into a BottomSheet component.
- We are not automatically changing our Modal to BottomSheet on code to avoid bundle size overhead as well as to avoid any flow to break unless mindfully implemented with BottomSheet on code.

### Modal for React Native
- Since we want to restrict the usage of Modals on native apps, we will not be implementing the Modal component for React Native.
- We will implement a dummy export which will only warn the consumer that this component is not supported on native apps.
- We will also add this to our documentation.

> ## Accessibility

Reviewer: Add a link to https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

- Trap keyboard focus within the modal.

> - Close the modal when the user hits the `esc` key.

Reviewer: If modal is opened via a trigger, focus should go back to the trigger after closing the modal.

Author: I think so. Are we doing the same on BottomSheet?

Reviewer: Yes.

- Add `aria-modal='true'`
- Add `aria-role='dialog'`

## Open questions

> - Should we have an upper limit on Modal stacking the same way we have an upper limit of 3 stacks on BottomSheet?

Reviewer: We do not have upper limit of bottomsheet stack of 3. We just render sheets on top of them as is. 

This upper limit was only discussed when Abhishek did the exploration of stacking where the stacking animation & UX worked like [Vercel's toast](https://vercel.com/design/toast)

Author: Got it

# Modal <!-- omit in toc -->

Modal components are often used to present information, notifications, or requests that require immediate attention or confirmation from the user. 

This document outlines the API of `Modal` component.

<!-- TODO: Add this -->
<!-- <img src="./modal-thumbnail.png" width="380" alt="Thumbnail of modal" /> -->

- [Design](#design)
- [Anatomy](#anatomy)
- [Animation](#animation)
- [API](#api)
  - [`Modal` API](#modal-api)
  - [`ModalBody` API](#modalbody-api)
  - [`ModalHeader` API](#modalheader-api)
  - [`ModalFooter` API](#modalfooter-api)
- [Behaviors](#behaviors)
  - [Modal Stacking](#modal-stacking)
  - [Modal Sizes \& Responsiveness](#modal-sizes--responsiveness)
  - [Modal Height](#modal-height)
  - [Modal Usage on mWeb \& native apps](#modal-usage-on-mweb--native-apps)
  - [Modal for React Native](#modal-for-react-native)
- [Accessibility](#accessibility)
- [Open questions](#open-questions)
- [References](#references)

## Design

- [Figma - Modal](https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?type=design&node-id=35860-607833&t=hEZsxB4yG6A3Gmib-0)

## Anatomy

> <!-- TODO: Add a better screenshot -->

Reviewer 2: lets also here 1 line description that the API/structure is similar to BottomSheet so people can get the context well ahead?? or better, place a screenshot of the bottomsheet side by side with anatomy

> <img src="./modal-anatomy.png" alt="Modal Anatomy" width="100%" />

Reviewer 2: is the image updated? the border-radius is 12px i think but on the animation the border radius is 8px if I'm not wrong. can we updated either of them to match the current design?

## Animation
**Entry:**
<img src="./modal-entry.gif" alt="Modal Entry Animation" width="100%" />

**Exit:**
<img src="./modal-exit.gif" alt="Modal Exit Animation" width="100%" />

## API

Sample usage:

```jsx
import { Modal, ModalHeader, ModalBody, ModalFooter, Button, Text } from '@razorpay/blade/components';

const [isOpen, setIsOpen] = useState(false);

> <Modal isOpen={isOpen} onDismiss={() => setIsOpen(false)}>

Reviewer 2: mention Modal will always be `controlled` component

Author: Adding it in the API section

  <ModalHeader title="Modal Title" />
  <ModalBody>
    <Text>This is the Modal Body</Text>
  </ModalBody>
  <ModalFooter>

>     <Button variant='secondary'>Cancel</Button>

Reviewer 2: the arrangement is in users hand right? we'll just render as is?

Author: Yes, the API specifies that Footer accepts JSX

    <Button variant='primary'>Save</Button>
  </ModalFooter>
</Modal>
```

### `Modal` API

| Prop            | Type                       | Default             | Description                                                                                                                                                                | Required |
| --------------- | -------------------------- | ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |
| isOpen          | `boolean`                  | `false`             | Toggles modal state                                                                                                                                                 |          |

> | size          | `small`, `medium`, `large`                  | `small`             | Size of the modal                                                                                                                                                 |          |

Reviewer 2: can you elaborate the description for `size` prop? so we can understand how will size translate to `*width` properties

| onDismiss       | `Callback`                 | `undefined`         | Called when the modal is closed, either by user state, hitting `esc` or tapping backdrop                                                                            |          |
| initialFocusRef | `React.Ref`                | `undefined`         | ref element you want to get keyboard focus when opening the modal                                                                                                          |          |

> | children | `React.ReactNode`                | `undefined`         | Accepts other Modal sub components like ModalHeader,ModalBody,ModalFooter                                                                                                          |          |

Reviewer 2: ```suggestion
| children | `React.ReactNode`                | `undefined`         | Accepts only Modal sub components like ModalHeader,ModalBody,ModalFooter                                                                                                          |          |
```

### `ModalBody` API

| Prop       | Type              | Default     | Description                 | Required |
| ---------- | ----------------- | ----------- | --------------------------- | -------- |
| `children` | `React.ReactNode` | `undefined` | Content of the Modal | ✅       |

> ### `ModalHeader` API

Author: As per design this was a conscious call not to have a back button on Modal. I will add this to open questions and get it checked with design.

| `title`   | `string`        | `undefined` | Title of the Header                   |       |
| `subtitle`   | `string`        | `undefined` | Subtitle of the Header                   |       |
| `leading` | `React.ReactNode` | `undefined` | leading asset or icon to be placed at the left most side of the ModalHeader |          |
| `trailing` | `Badge, Link, Text, IconButton`, `undefined` | undefined  |     trailing component to be placed at the right most side of the ModalHeader     |    |
| `titleSuffix` | `Counter` | `undefined` | A component to be placed adjacent to the title text |          |

### `ModalFooter` API
> `ModalFooter` API will be similar to [`BottomSheetFooter` API](https://github.com/razorpay/blade/blob/master/packages/blade/src/components/BottomSheet/_decisions/decisions.md#bottomsheetfooter-api).

| Prop       | Type              | Default     | Description                 | Required |
| ---------- | ----------------- | ----------- | --------------------------- | -------- |
| `children` | `React.ReactNode` | `undefined` | Contents of the ModalFooter | ✅       |

## Behaviors

> ### Modal Stacking

Reviewer 2: we shall also have a section for Modal Closing and then add things like
- Close on Esc(although we have this in a11y) but not sure if that's relevant section
- Similarly, we clicking on Overlay outside Modal shall also close modal? If yes, we shall document that here

- We're at consensus that we do need to support modal stacking.
- Design is not yet aligned on how the modal should look like when stacked and what the motion should be.
- As an interim solution we can implement a basic stacking behavior with the same exit/entry animations as the current modal and look into enhancing this motion later.

> ### Modal Sizes & Responsiveness

Reviewer 2: this section also has a lot of content. Preferable would be better to put more screenshots and less words that will set the context easily. This whole logic is very confusing when broken down into a lot of points

- We will offer Modal in 3 sizes - small, medium and large.
- Each size corresponds to a max-width for the Modal
- As of today these are the max-widths we're using for each size:
  - small: `400px`
  - medium: `760px`
  - large: `1024px`
- All the Modals will have a min-width of `320px` regardless of the size.
- This would mean that the Modal will be responsive and will scale down to `320px` on smaller screens but will not scale up beyond the max-width as per its size mentioned above.

> - POC Link: https://codesandbox.io/s/eloquent-lalande-jvxwno

Reviewer 2: opened the POC both look same, can't spot difference when resizing the window too 🤔

Author: Yup, both logic work

### Modal Height

> - Each Modal will have a max-height of `80vh` (80% of viewport height) and will be scrollable if the content exceeds this height.

Reviewer 2: what will be the default height if the content is smaller than 80vh?

Author: The content's height

> ### Modal Usage on mWeb & native apps

Reviewer 2: all these sections have too many points. Did chatgpt wrote it for you? seems repetitive and have overlaps. can we simplify this and combine the points so we can remove unnecessary points? Like for example

```
- We will not be using the Modal component on mWeb or native apps.
- We will only use the BottomSheet component on mWeb and native apps.
```

could only be one single point
```
- We will not be exposing the Modal component on mWeb or native apps(this is a design guideline) and as a preference BottmSheet shall be use on mobile platforms(mweb and RN)
```

> - We will not be using the Modal component on mWeb or native apps.

Reviewer 2: ```suggestion
- We will not be exposing the Modal component on mWeb or native apps(this is a design guideline).
```

- We will only use the BottomSheet component on mWeb and native apps.
- Our Modal will be responsive and will not break up till 320px on mWeb but will not be usable on native apps.
- We will throw a warning when a Modal is opened on smaller screens.
- Designers would need to be mindful of this while designing for mWeb and native apps.
- On Figma, we have added guardrails to ensure that when screenSize is selected Mobile on Figma, Modal component will turn into a BottomSheet component.

> - We are not automatically changing our Modal to BottomSheet on code to avoid bundle size overhead as well as to avoid any flow to break unless mindfully implemented with BottomSheet on code.

Reviewer 2: ```suggestion
- We are **not automatically** changing our Modal to BottomSheet on code to avoid bundle size overhead as well as to avoid any flow to break unless mindfully implemented with BottomSheet on code.
```
make it bold 😅

Reviewer 2: Also we can back link this to Bottomsheet - I like this section a lot. it says in my face about reasonings with screenshot. probably we can steal from there and add it here as well https://github.com/razorpay/blade/blob/master/packages/blade/src/components/BottomSheet/_decisions/decisions.md#composition-with-dropdown

### Modal for React Native
- Since we want to restrict the usage of Modals on native apps, we will not be implementing the Modal component for React Native.
- We will implement a dummy export which will only warn the consumer that this component is not supported on native apps.
- We will also add this to our documentation.

## Accessibility
- Trap keyboard focus within the modal.
- Close the modal when the user hits the `esc` key.
- Add `aria-modal='true'`
- Add `aria-role='dialog'`
- Expose `accessibilityLabel` prop

> - Ref: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

Reviewer 2: add the page scroll-lock behavior when the modal will be in opened state?

## Open questions

> - Do we need a different animation for Modal stacking? Yet to explore on design

Reviewer 2: are we accommodating full screen modal's use cases?

Author: We don't have any use-cases for full screen modals. Let me check with design.

## Animation
**Entry:**
<img src="./modal-entry.gif" alt="Modal Entry Animation" width="100%" />

**Exit:**
<img src="./modal-exit.gif" alt="Modal Exit Animation" width="100%" />

## API

Sample usage:

```jsx
import { Modal, ModalHeader, ModalBody, ModalFooter, Button, Text } from '@razorpay/blade/components';

const [isOpen, setIsOpen] = useState(false);

<Modal isOpen={isOpen} onDismiss={() => setIsOpen(false)}>
  <ModalHeader title="Modal Title" />
  <ModalBody>
    <Text>This is the Modal Body</Text>
  </ModalBody>
  <ModalFooter>
    <Button variant='secondary'>Cancel</Button>
    <Button variant='primary'>Save</Button>
  </ModalFooter>
</Modal>
```

### `Modal` API

| Prop            | Type                       | Default             | Description                                                                                                                                                                | Required |
| --------------- | -------------------------- | ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |

> | isOpen          | `boolean`                  | `false`             | Toggles modal state                                                                                                                                                 |          |

Author: Yeah, makes more sense to mark it required. It would be unusable unless they handle isOpen and onDismiss

- Each of these actions will call the `onDismiss` callback passed by the consumer.

### Modal Stacking
- We will be supporting stacking of Modals.
- As an initial implementation, we will have a basic stacking animation where the new Modal will have its own default entry and exit animation without affecting the existing Modal.
- As an enhancement post release, we will explore if we can have a more sophisticated stacking animation.

### Modal Sizes & Responsiveness

<img src="./modal-responsiveness.gif" alt="Modal Responsiveness POC" width="100%" />

- We will offer Modal in 3 sizes - small, medium and large.
- Each size corresponds to a `max-width` for the Modal
- As of today these are the max-widths we're using for each size:
  - small: `400px`
  - medium: `760px`
  - large: `1024px`
- All the Modals will have a `width` of `100vw - 32px` (16px gap on left & right sides) so that there is always a gap between the Modal and the viewport.
- All the sizes of Modals will have a `min-width` of `320px`.
  - This would mean that the Modal will be responsive and will scale down to `320px` on smaller screens but will not scale up beyond the max-width as per its size mentioned above.

> - **POC Link**: https://codesandbox.io/s/eloquent-lalande-jvxwno

Author: Yup, max width will be based on sizes as mentioned in the points above

- Even though usage on mWeb is discouraged by our design guideline, our Modal will be responsive and will not visually break till 320px on mWeb.
- We will throw a warning when a Modal is opened on smaller screens as well as add this to our documentation.
- Designers would need to be mindful of this while designing for mWeb and native apps.
- On Figma, we have added guardrails to ensure that when screenSize is selected Mobile on Figma, Modal component will turn into a BottomSheet component.
- We are **not automatically changing** our **Modal to BottomSheet** on code to avoid bundle size overhead as well as to avoid any flow to break unless mindfully implemented with BottomSheet on code.

#### Using BottomSheet on mWeb
- As per our design guideline, we will be using BottomSheet on mWeb for all the use-cases where we would have used Modal on dWeb.
- This would require our consumers to add some logic to handle this switch from Modal to BottomSheet on mWeb.
- This decision also enables our consumers to save on bundle size so that the mWeb's bundle only contains BottomSheet and the dWeb's bundle only contains Modal.

**Example usage:**
```jsx

> import { useTheme, BottomSheet, Modal } from '@razorpay/blade';

Reviewer 3: importing both bottomsheet and modal will anyway inflate bundle for both platforms, we should recommend lazy loading them (likely with Suspense)

Author: Yeah. I will put up a note here

| `children` | `React.ReactNode` | `undefined` | Content of the Modal | ✅       |

### `ModalHeader` API
> `ModalHeader` API will be similar to [`BottomSheetHeader` API](https://github.com/razorpay/blade/blob/master/packages/blade/src/components/BottomSheet/_decisions/decisions.md#bottomsheetheader-api) with the exception of `showBackButton` & `onBackButtonClick` props.

| Prop      | Type            | Default     | Description                           | Required |
| --------- | --------------- | ----------- | ------------------------------------- | -------- |
| `title`   | `string`        | `undefined` | Title of the Header                   |       |
| `subtitle`   | `string`        | `undefined` | Subtitle of the Header                   |       |
| `leading` | `React.ReactNode` | `undefined` | leading asset or icon to be placed at the left most side of the ModalHeader |          |
| `trailing` | `Badge, Link, Text, IconButton`, `undefined` | undefined  |     trailing component to be placed at the right most side of the ModalHeader     |    |
| `titleSuffix` | `Counter` | `undefined` | A component to be placed adjacent to the title text |          |

> ### `ModalFooter` API

Author: Yup its an optional component

- Designers would need to be mindful of this while designing for mWeb and native apps.
- On Figma, we have added guardrails to ensure that when screenSize is selected Mobile on Figma, Modal component will turn into a BottomSheet component.
- We are **not automatically changing** our **Modal to BottomSheet** on code to avoid bundle size overhead as well as to avoid any flow to break unless mindfully implemented with BottomSheet on code.

#### Using BottomSheet on mWeb
- As per our design guideline, we will be using BottomSheet on mWeb for all the use-cases where we would have used Modal on dWeb.
- This would require our consumers to add some logic to handle this switch from Modal to BottomSheet on mWeb.
- This decision also enables our consumers to save on bundle size so that the mWeb's bundle only contains BottomSheet and the dWeb's bundle only contains Modal.

**Example usage:**

> You can use Lazy Loading to load the Modal component only on dWeb and BottomSheet component only on mWeb.

```jsx
import { useTheme, BottomSheet, Modal } from '@razorpay/blade';

const App = () => {

>   const { theme, platform } = useTheme();

Author: Yup blade provides `useTheme` I see its not documented as well as it should be. I will create an issue for this.

  return (
    <>
      {platform === 'onMobile' ? (
        <BottomSheet>
          <BottomSheetHeader />
          <BottomSheetBody>
            /* BottomSheetBody content */
          </BottomSheetBody>
          <BottomSheetFooter />
        </BottomSheet>
      ) : (
        <Modal>
          <ModalHeader />
          <ModalBody>
            /* ModalBody content */
          </ModalBody>
          <ModalFooter />
        </Modal>
      )}
    </>
  );
};
```

### Modal for React Native

> - Since we want to restrict the usage of Modals on native apps, we will not be implementing the Modal component for React Native.

Reviewer 4: @chaitanyadeorukhkar For mobile we will be using only bottomsheet going forward?

Author: Yeah we will only be using bottomsheet for all use-cases
